shims/spark310/src/main/scala/com/nvidia/spark/rapids/shims/spark310/ParquetCachedBatchSerializer.scala

```diff
@@ -231,27 +345,139 @@ class ParquetCachedBatchSerializer extends CachedBatchSerializer with Arm {
         new ColumnarBatch(cols.map(_.copyToHost()).toArray, gpuBatch.numRows())
       }
     })
-    cbRdd.mapPartitions(iter => new CloseableColumnBatchIterator(iter))
+    cbRdd.mapPartitions(iter => CloseableColumnBatchIterator(iter))
   }
 
   /**
    * Convert the cached batch into `InternalRow`s.
+   *
    * @param input the cached batches that should be converted.
    * @param cacheAttributes the attributes of the data in the batch.
    * @param selectedAttributes the field that should be loaded from the data and the order they
    *                           should appear in the output rows.
    * @param conf the configuration for the job.
    * @return RDD of the rows that were stored in the cached batches.
    */
-  override def convertCachedBatchToInternalRow(input: RDD[CachedBatch],
-                                cacheAttributes: Seq[Attribute],
-                                selectedAttributes: Seq[Attribute],
-                                conf: SQLConf): RDD[InternalRow] = {
-    val cb = convertCachedBatchToColumnarBatch(input, cacheAttributes, selectedAttributes, conf)
-    val rowRdd = cb.mapPartitions(iter => {
-      new ColumnarToRowIterator(iter)
-    })
-    rowRdd
+  override def convertCachedBatchToInternalRow(
+     input: RDD[CachedBatch],
+     cacheAttributes: Seq[Attribute],
+     selectedAttributes: Seq[Attribute],
+     conf: SQLConf): RDD[InternalRow] = {
+
+    val sparkSession = SparkSession.active
+    val hadoopConf = getHadoopConf(selectedAttributes.toStructType, conf)
+    val broadcastedHadoopConf =
+      sparkSession.sparkContext.broadcast(new SerializableConfiguration(hadoopConf))
+
+    input.mapPartitions {
+      cbIter => {
+        new Iterator[InternalRow]() {
+          var iter: Iterator[InternalRow] = null
+
+          override def hasNext: Boolean =
+            // either the current iterator has an element,
+            // else go over the batch and get the next non-degenerate iterator
+            (iter != null && iter.hasNext) || doesNextValidIteratorHasNext
```

Comment:
Once the while loop ordering is fixed, I think we can just fold `doesNextValidIteratorHasNext` directly into this method and not need to separate it out.  The logic is small enough to be easily understood in one place rather than splitting it across methods.